libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/TripServiceImpl.kt

```diff
@@ -0,0 +1,145 @@
+package com.mapbox.services.android.navigation.v5.internal.navigation
+
+import android.app.Service
+import android.content.Intent
+import android.os.IBinder
+import com.mapbox.android.core.location.LocationEngine
+import com.mapbox.android.core.location.LocationEngineCallback
+import com.mapbox.android.core.location.LocationEngineRequest
+import com.mapbox.android.core.location.LocationEngineResult
+import com.mapbox.navigation.base.trip.NavigationTripDescriptor
+import com.mapbox.navigation.base.trip.TripNotification
+import com.mapbox.navigation.base.trip.TripService
+import com.mapbox.navigation.navigator.MapboxNativeNavigatorImpl
+import com.mapbox.navigation.utils.extensions.ifNonNull
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+const val NAVIGATION_SERVICE_ID = 0xff00aa
+const val UPDATE_INTERVAL_IN_MILLISECONDS: Long = 1000
+const val FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS: Long = 500
+
+@ExperimentalCoroutinesApi
+internal object NavigationNotificationService : Service() {
+    private val job = SupervisorJob()
+    val serviceScope = CoroutineScope(job + Dispatchers.IO)
+
+    var locationEngineCallback = object : LocationEngineCallback<LocationEngineResult> {
+        override fun onSuccess(result: LocationEngineResult?) {
+        }
+
+        override fun onFailure(exception: java.lang.Exception) {
+        }
+    }
+    var locationEngine: LocationEngine? = null
+    val locationEngineRequest: LocationEngineRequest = LocationEngineRequest.Builder(UPDATE_INTERVAL_IN_MILLISECONDS)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .setFastestInterval(FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS)
+            .build()
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return null
+    }
+
+    override fun onCreate() {
+        startForeground(NAVIGATION_SERVICE_ID, null)
+        notificationChannel.close()
+        notificationChannel = ConflatedBroadcastChannel()
+        serviceScope.launch {
+            while (isActive) {
+                val status = MapboxNativeNavigatorImpl.getStatus(java.util.Date()) // Request that will drive the Polling functionality
+                locationEngine?.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {
```

Comment:
Do we want to handle last location updates this way? Noticed that `locationEngineCallback` is empty and wondering if we should keep track of the last location received in `locationEngineCallback` (`requestLocationUpdates`) and use it here so we don't need to create the location engine callback object here. I believe you were trying to avoid mutability, right? What are other advantages of implementing it like this?

Reply:
This piece of code is removed.